temoto_context_manager/include/temoto_context_manager/env_model_repository.h

```diff
@@ -0,0 +1,199 @@
+#ifndef ENV_MODEL_REPOSITORY_H
+#define ENV_MODEL_REPOSITORY_H
+
+namespace temoto_context_manager 
+{
+namespace emr 
+{
+
+
+
+/**
+ * @brief Abstract base class for payloads
+ * 
+ */
+class PayloadEntry
+{
+private:
+  
+public:
+  std::string type;
+
+  PayloadEntry(std::string type) : type(type) {}
+
+  ~PayloadEntry() {}
+
+  PayloadEntry() {}
+
+  virtual std::string getName() = 0;
+  /**
+   * @brief Get the type of the Payload
+   * 
+   * @return std::string 
+   */
+  std::string getType() {return type;}
+};
+
+template <class ROSMsg>
+class ROSPayload : public PayloadEntry
+{
+private:
+  ROSMsg payload_;
+public:
+  /**
+   * @brief Get the Name object
+   * 
+   * @return std::string 
+   */
+  std::string getName()
+  {
+    return payload_.name;
+  }
+  ROSMsg getPayload() {return payload_;};
+  /**
+   * @brief Set the Payload object
+   * 
+   * @param payload 
+   */
+  void setPayload(ROSMsg & payload) {payload_ = payload;};
+
+  ROSPayload(ROSMsg payload) : payload_(payload)
+  {
+  }
+
+};
+
+/**
+ * @brief A single node in the EMR tree
+ * 
+ * A node contains a payload and pointers to its (singular) parent and children.
+ * 
+ */
+class Node : public std::enable_shared_from_this<Node>
+{
+private:
+  std::weak_ptr<Node> parent_;
+  std::vector<std::shared_ptr<Node>> children_;
+  std::shared_ptr<PayloadEntry> payload_;
+
+public:
+
+  void addChild(std::shared_ptr<Node> child);
+  /**
+   * @brief Set the Parent pointer
+   * 
+   * NB! Don't use this manually. Use addChild() of parent instead.
+   * 
+   * @param parent 
+   */
+  void setParent(std::shared_ptr<Node> parent);
+  std::weak_ptr<Node> getParent()
+  {
+    return parent_;
+  }
+  /**
+   * @brief Get children of node
+   * 
+   * @return std::vector<std::shared_ptr<Node>> 
+   */
+  std::vector<std::shared_ptr<Node>> getChildren()
```

Comment:
Just a heads up that if the "children" are only meant to be read (are not meant to be modified), then it is better to return a const reference of the thing you are returning (const std::vector<std::shared_ptr<Node>>& getChildren())

Reply:
Noted and refactored.